framework/src/st_things/things_stack/utils/things_network.c

```diff
@@ -259,31 +243,57 @@ int things_wifi_changed_call_func(int state, char *ap_name, char *ip_addr)
 
 static void *__attribute__((optimize("O0"))) t_things_wifi_join_loop(void *args)
 {
+	bool* is_wifi_retry_connect = (bool*) args;
+
+	if ((*is_wifi_retry_connect) == true) {
+		THINGS_LOG_E(TAG, "Retry to connect to home ap.");
+		things_network_connect_home_ap();
+	} else {
 #ifdef CONFIG_NETUTILS_NTPCLIENT
-	sync_time_from_ntp(10);
+		sync_time_from_ntp(10);
 #else
-	THINGS_LOG_V(TAG, "CONFIG_NETUTILS_NTPCLIENT is not set");
+		THINGS_LOG_V(TAG, "CONFIG_NETUTILS_NTPCLIENT is not set");
 #endif
+		wifi_manager_info_s wifi_info;
+		wifi_manager_get_info(&wifi_info);
 
-	wifi_manager_info_s wifi_info;
-	wifi_manager_get_info(&wifi_info);
-
-	things_wifi_changed_call_func(1, wifi_info.ssid, wifi_info.ip4_address);
+		things_wifi_changed_call_func(1, wifi_info.ssid, wifi_info.ip4_address);
+	}
 
 	return NULL;
 }
 
 void things_wifi_sta_connected(wifi_manager_result_e res)
 {
+	bool is_wifi_retry_connect = false;
+
 	if (res == WIFI_MANAGER_FAIL) {
 		THINGS_LOG_E(TAG, "Failed to connect to the AP");
-		return;
+
+		if (!dm_is_es_complete()) {
+			g_retry_connect_cnt++;
+
+			if (g_retry_connect_cnt <= MAX_WIFI_CONNECT_RETRY_COUNT) {
+				is_wifi_retry_connect = true;
+				THINGS_LOG_V(TAG, "(%d) Retry to connect ap during easy-setup.", g_retry_connect_cnt);
+			} else {
+				return;
+			}
+		} else {
+			is_wifi_retry_connect = true;
+			THINGS_LOG_V(TAG, "Retry to connect ap during normal time.");
+		}
+	} else {
+		is_wifi_retry_connect = false;
```

Comment:
easysetup is always success if wifi manager returns CB_STA_CONNECTED?

Reply:
dm_is_es_complete means the all steps of easy-setup was done.
While running the easy-setup, the codes should retry only 3 times then has to return a fail because the mobile can receive about failured message.

Reply:
I mean if wifi manager returns,  CB_STA_CONNECTED, it will comes here, else part.
In this case CB_STA_CONNECTED guarantees easy  setup completed??That's what I curious

Reply:
dm_is_es_complete guarantees thing which easy-setup was done.
And the things should be connected some AP. 

The things was trying to connect during the easy-setup then it should return any error, if cannot connect. so it tries only 3 times. But, after easy-setup was done, tries to connect infinites.

Sure, it has some dependency for the user scenario.